### PR TITLE
15 train baseline model

### LINKED ISSUE #15: Train baseline model

### PR DESCRIPTION
**Pull request by RFOxbury**
---

# Description

Contains code for two potential baseline models: one based on label probabilities, and one that just predicts the most common combination of labels each time.

Fixes #15 

# Instructions for Reviewer

In order to test the code in this PR you need to ...

`analysis/baseline_model.py` and `notebooks/04_train_baseline_model` overlap - the former is a script that can easily be called from the command line and that logs the model on weights and biases; the latter is a notebook that includes EDA and compares metrics between the two baseline models, and also between training and validation sets.

- Read and run through `notebooks/04_train_baseline_model.py`
- Run `baseline_model.py` from the command line and check that it runs. There are instructions how to do this at the top of the script.
- Run unit tests via `pytest discovery_child_development/tests` and they should all pass

Please pay special attention to ...

- Is `discovery_child_development/analysis/` the right place for `baseline_model.py` or should it go in `.../pipeline/`?

# Checklist:

- [x] I have refactored my code out from `notebooks/`
- [x] I have checked the code runs
- [x] I have tested the code
- [x] I have run `pre-commit` and addressed any issues not automatically fixed
- [x] I have merged any new changes from `dev`
- [x] I have documented the code
  - [x] Major functions have docstrings
  - [ ] Appropriate information has been added to `README`s
- [x] I have explained this PR above
- [x] I have requested a code review

### CODE REVIEW

File: discovery_child_development/analysis/baseline_model.py [hunk 1]
```diff
@@ -0,0 +1,423 @@
...
+load_dotenv()
+
+S3_BUCKET = os.environ.get("S3_BUCKET")
+PARAMS = import_config("config.yaml")
+CONCEPT_IDS = "|".join(PARAMS["openalex_concepts"])
+INPUT_PATH = "data/openAlex/processed/"
+VECTORS_FILEPATH = "data/openAlex/vectors/sentence_vectors_384.parquet"
+DATA_PATH_LOCAL = os.path.join(PROJECT_DIR, "inputs", "data")
+FIG_PATH = os.path.join(PROJECT_DIR, "outputs", "figures")
+MODEL_PATH = os.path.join(PROJECT_DIR, "outputs", "models")
```
Comment (beingkk) [inline discovery_child_development/analysis/baseline_model.py]: I think the following notation looks neater - but it's a matter of preference:
```python
DATA_PATH_LOCAL = PROJECT_DIR / "inputs/data" 
```

File: discovery_child_development/analysis/baseline_model.py [hunk 2]
```diff
@@ -0,0 +1,423 @@
...
+class MostCommonClassifier(BaseEstimator, ClassifierMixin):
+    """
+    This class implements a classifier that predicts the same combination of labels for every single new datapoint.
+    The combination of labels provided during initialization is used as the prediction for all input samples.
+
+    Attributes:
+        labels (Iterable): The combination of labels that will be predicted for each new datapoint.
+    """
+
+    def __init__(self, labels):
```
Comment (beingkk) [inline discovery_child_development/analysis/baseline_model.py]: I'd suggest adding types also in function definitions, not only the docstrings 🙏 

Applies here and elsewhere in this file

File: discovery_child_development/analysis/baseline_model.py [hunk 3]
```diff
@@ -0,0 +1,423 @@
...
+
+def find_most_frequent_labels(
+    df: pd.DataFrame, label_col: str = "sub_category", head: int = 20
+) -> Tuple[pd.Series, Union[Tuple, List]]:
+    """
+    Finds the most frequent combinations of labels in a specified column of a dataframe, where each entry in the column
+    contains a list or tuple of labels. The function returns the top combinations and the most frequent combination of labels.
+
+    Args:
+        df (pd.DataFrame): The dataframe containing the label data.
```
Comment (beingkk) [inline discovery_child_development/analysis/baseline_model.py]: Can you please indicate what kind of data/columns does the label dataframe need to have?

File: discovery_child_development/analysis/baseline_model.py [hunk 4]
```diff
@@ -0,0 +1,423 @@
...
+
+    if model_type not in valid_inputs:
+        raise ValueError(
+            f"Invalid input. Expected one of {valid_inputs}, got '{model_type}'"
+        )
+
+    # Load the data
+    training_data_filename = (
+        f"openalex_data_{concept_ids}_year-2019-2020-2021-2022-2023_train.csv"
+    )
```
Comment (beingkk) [inline discovery_child_development/analysis/baseline_model.py]: Perhaps better to define this as a global variable at the top, as we eventually might want to use a training dataset with a different filename pattern? (eg, if we not only use concept_ids but also keyword search to expand our OpenAlex subset - or if we combine our patent and OpenAlex datasets)

File: discovery_child_development/analysis/baseline_model.py [hunk 5]
```diff
@@ -0,0 +1,423 @@
...
+
+    embeddings = embeddings.set_index("openalex_id")
+
+    openalex_data_wide = openalex_data_wide.join(
+        embeddings, on="openalex_id", how="left"
+    )
+
+    Y, mlb = classification_utils.add_binarise_labels(
+        openalex_data_wide, label_column="sub_category", not_valid_label=None
+    )
```
Comment (beingkk) [inline discovery_child_development/analysis/baseline_model.py]: Apologies, I'm not clear here - is your label here a single sub-category or the combination of all sub-categories?

I think the dataset (openalex_data_wide) before "binarisation" generally should be instances of single document ID - sub-category pairs. So, if say paper_1 is tagged with category1 and category2 then I'd expect two rows - as opposed to just on row with the label being (category1, category2).

File: discovery_child_development/analysis/baseline_model.py [hunk 6]
```diff
@@ -0,0 +1,423 @@
...
+
+    train_ids, val_ids = train_test_split(unique_ids, test_size=0.1, random_state=SEED)
+
+    X_train = (
+        openalex_data_wide[openalex_data_wide.index.isin(train_ids)][
+            "miniLM_384_vector"
+        ]
+        .apply(pd.Series)
+        .values
+    )
```
Comment (beingkk) [inline discovery_child_development/analysis/baseline_model.py]: I guess for the baseline models you didn't really need embeddings - but you've kept as it will be useful further down the line?

File: discovery_child_development/analysis/baseline_model.py [hunk 7]
```diff
@@ -0,0 +1,423 @@
...
+    # So drop_duplicates ensures that each sub-category is only counted once per OpenAlex ID
+    df_cleaned = df.drop_duplicates(subset=["openalex_id", label_col])
+    label_probabilities = pd.DataFrame(df_cleaned[label_col].value_counts())
+    # Normalise using the number of unique documents as the denominator (each label can appear at most once per document because of
+    # how we cleaned the data above, so we have made sure that the max "prob" possible is 1)
+    label_probabilities["prob"] = label_probabilities[label_col] / n
+    label_probabilities = label_probabilities[
+        ["prob"]
+    ]  # drop the count column, so it is just index and probability
+    label_probabilities = label_probabilities.squeeze()  # convert to a Series
```
Comment (beingkk) [inline discovery_child_development/analysis/baseline_model.py]: Thanks, I'd suggest chaining the panda functions so that it's easier to read:
```
label_probabilities = (
   pd.DataFrame(...
   .assign(prob = ...)
    ...
)
```

File: discovery_child_development/analysis/baseline_model.py [hunk 8]
```diff
@@ -0,0 +1,423 @@
...
+def get_label_probabilities(
+    df: pd.DataFrame, label_col: str = "sub_category", n: int = 1000, targets=None
+) -> pd.Series:
+    """Get label probabilities from long-form data
+
+    Args:
+        df (pd.DataFrame): a long-form dataframe with an ID column and a label column (where each ID can have multiple labels)
+        label_col (str, optional): The column that contains the label. Defaults to "sub_category".
+        n (int, optional): Denominator used when calculating the prevalence of labels. Should be the number of unique IDs. Defaults to 1000.
+        targets (_type_, optional): This should be used if you're calculating probabilities on the training set, but the training set does not contain all possible labels. Defaults to None.
```
Comment (beingkk) [inline discovery_child_development/analysis/baseline_model.py]: What exactly is this - labels that you definitely want to appear?

File: discovery_child_development/analysis/baseline_model.py [hunk 9]
```diff
@@ -0,0 +1,423 @@
...
+        .groupby(["openalex_id", "text"])["sub_category"]
+        .agg(lambda x: tuple(set(x)))
+        .reset_index()
+    )
+    # Set the index - useful later for creating training/validation split
+    openalex_data_wide = openalex_data_wide.set_index("openalex_id")
+
+    top_combinations, most_common_combination = find_most_frequent_labels(
+        openalex_data_wide, label_col="sub_category", head=20
+    )
```
Comment (beingkk) [inline discovery_child_development/analysis/baseline_model.py]: Are you using the `most_common_combination` variable here later on in the function - if not, can replace with `_`

Also, I'd call this function where it is needed, ie after you've checked `if model_type == "majority_combination":`

File: discovery_child_development/analysis/baseline_model.py [hunk 10]
```diff
@@ -0,0 +1,423 @@
...
+
+    Y, mlb = classification_utils.add_binarise_labels(
+        openalex_data_wide, label_column="sub_category", not_valid_label=None
+    )
+
+    # Split IDs into random train and test subsets
+    unique_ids = openalex_data_wide.index.unique()
+
+    train_ids, val_ids = train_test_split(unique_ids, test_size=0.1, random_state=SEED)
```
Comment (beingkk) [inline discovery_child_development/analysis/baseline_model.py]: Just noting that you're splitting your dataset by IDs - I wonder what are the pros/cons for splitting not by IDs but by ID-label pairs (as you can have more than one label per ID)

Here, however, I guess we come back to the point about how you've encoded the labels. Again, I'd suggest encoding the labels not as combinations of sub-categories but simply each label as a single sub-category (see previous comment)

**Review by beingkk — CHANGES_REQUESTED** [review]
Thank you for this - great to see that we have a couple basic baselines to compare the model training with!

I managed to successfully run the code - and the tests passed as well.

I had some minor comments and one main clarifying question about the way you encode the labels (ie, whether a label is a sub-category or the combination of sub-categories for each paper ID). I think we should encode the labels such that each label is one sub-category rather than combinations.

Best wishes!

**Comment by beingkk** [comment]
Oh, regarding your last question - I think we can move the baseline model script into pipeline, as it will be part of the model training and checking pipeline. Perhaps we can create a separate subfolder for 'models' ?

The naming of the 'analysis' folder doesn't make that much sense to be fair, if we already have a notebooks folder.

File: discovery_child_development/analysis/baseline_model.py (hunk 8 above)
Comment (RFOxbury) [reply discovery_child_development/analysis/baseline_model.py]: Hey, now that I look at it, this function is kind of messy, so let me know if you want me to rethink the process a bit, rather than just adding more info in the docstring.

The way this function is used in context is:
```
label_probabilities = get_label_probabilities(
            train_df,
            "sub_category",
            len(train_df["openalex_id"].unique()),
            targets=Y_train.columns,
        )
```
In this case, `train_df` was a long-form dataframe (one row per label) and did not contain every possible label. `Y` was created with a multilabel binarizer and later split into `Y_train` and `Y_test`, hence `Y_train` contains a column for every possible label.

I don't know if the process is robust... Is it data leakage if we're artificially adding labels from the test set to the training set like this? (Not that leakage is an issue for this kind of model anyway, but once it's a "real" model it would be a problem)

File: discovery_child_development/analysis/baseline_model.py (hunk 4 above)
Comment (RFOxbury) [reply discovery_child_development/analysis/baseline_model.py]: Good spot, thank you!

File: discovery_child_development/analysis/baseline_model.py (hunk 9 above)
Comment (RFOxbury) [reply discovery_child_development/analysis/baseline_model.py]: 🤯 thanks, yes I can use _ instead!

And good spot - it is indeed only needed for the "majority_combination" type

File: discovery_child_development/analysis/baseline_model.py (hunk 5 above)
Comment (RFOxbury) [reply discovery_child_development/analysis/baseline_model.py]: I actually got that `add_binarise_labels` function from the other Innovation Sweet Spots repo [here](https://github.com/nestauk/innovation_sweet_spots/blob/dev/innovation_sweet_spots/analysis/notebooks/review_labelling/utils.py). I think that function expects a dataframe where there is one row per datapoint (per openalex ID) and the label column contains a list (or tuple) of labels

File: discovery_child_development/analysis/baseline_model.py (hunk 10 above)
Comment (RFOxbury) [reply discovery_child_development/analysis/baseline_model.py]: That's a really good point. My first thought is that if the labels are correlated in some way (eg documents with "label 1" are also likely to be tagged with "label 2", whereas documents with "label 3" are likely to have no other labels) it would be important to make sure that no IDs straddle the training and validation sets, so that the model doesn't already know that an ID in the validation set has been tagged with "label 1" (and therefore has a high probability of also having "label 2").

File: discovery_child_development/analysis/baseline_model.py (hunk 6 above)
Comment (RFOxbury) [reply discovery_child_development/analysis/baseline_model.py]: Yes... I didn't need to load them at all 😅 Not sure why I did tbh! I guess it's useful for imagining what the eventual pipeline might look like but it's not necessary here.

File: discovery_child_development/analysis/baseline_model.py (hunk 5 above)
Comment (beingkk) [reply discovery_child_development/analysis/baseline_model.py]: Great, apologies for the confusion!

File: discovery_child_development/analysis/baseline_model.py (hunk 8 above)
Comment (beingkk) [reply discovery_child_development/analysis/baseline_model.py]: I see - I guess you could only use `Y_train` instead of `train_df` - but it's probably ok for now!

And I think it's OK to define every possible label beforehand, and this is not data leakage.

File: discovery_child_development/analysis/baseline_model.py (hunk 10 above)
Comment (beingkk) [reply discovery_child_development/analysis/baseline_model.py]: Yeah, I guess let's keep the IDs separate across test and train sets as you've done so far.

File: discovery_child_development/analysis/baseline_model.py (hunk 1 above)
Comment (RFOxbury) [reply discovery_child_development/analysis/baseline_model.py]: Thanks! Coincidentally I've also recently seen this mentioned in the [code review guidance](https://github.com/nestauk/github_support/blob/dev/guidelines/reviewing_pr.md) I have changed this accordingly
